File: solutions/Best-Time-to-Buy-and-Sell-Stock-III/Best-Time-to-Buy-and-Sell-Stock-III.cpp

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int n=prices.size();
        int ans=0;
        if(n==0)
            return ans;
        vector<int> l(n,0);
        vector<int> r(n,0);
        int lastmin=prices[0];
        for(int i=1;i<n;i++)
        {
            lastmin=min(lastmin,prices[i]);
            l[i]=max(l[i-1],prices[i]-lastmin);
        }
        int lastmax=prices[n-1];
        for(int i=n-2;i>=0;i--)
        {
            lastmax=max(lastmax,prices[i]);
            r[i]=max(r[i+1],lastmax-prices[i]);
            ans=max(ans,l[i]+r[i]);
        }
        return ans;
    }
};


 

File: solutions/Best-Time-to-Buy-and-Sell-Stock-III/Best-Time-to-Buy-and-Sell-Stock-III.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let hold1 = Number.NEGATIVE_INFINITY;
  let hold2 = Number.NEGATIVE_INFINITY;
  let release1 = 0;
  let release2 = 0;
  const length = prices.length;
  for (let i = 0; i < length; i++) {
    release2 = Math.max(release2, hold2 + prices[i]);
    hold2 = Math.max(hold2, release1 - prices[i]);
    release1 = Math.max(release1, hold1 + prices[i]);
    hold1 = Math.max(hold1, -prices[i]);
  }
  return release2;

};